Added documentation (README) for all functionalities (#58)
NBISweden/sda-cli: README.md

This tool can be used to encrypt and upload data when submitting to the archive,
and to download and decrypt with retrieving data from the archive.

To get help on the usage of the tool, please use the following command
```bash
./sda-cli help

> ``` 

Reviewer: ```suggestion
```
```

# Installation

> This section contains the information required to install, modify and run the `sda-cli` tool. 

Reviewer: ```suggestion
This section contains the information required to install, modify and run the `sda-cli` tool.
```

# Requirements
The `sda-cli` is written in golang. In order to be able to modify, build and run the tool, golang needs to be installed. The instructions for installing go can be found [here](https://go.dev/doc/install).

### Build tool
To build the `sda-cli` tool run the following command from the root folder of the repository
```bash
go build
```
This command will create an executable file in the root folder, named `sda-cli`.

# Usage
The main functionalities implemented in this tool are explained in the following sections.

## Encrypt

The files stored in the SDA/BP archive are encrypted using the [crypt4gh standard](https://www.ga4gh.org/news/crypt4gh-a-secure-method-for-sharing-human-genetic-data/). The following sections explain how to encrypt and upload files to the archive.

### Download the crypt4gh public key

> The files that are uploaded to the SDA/BP services, need to be encrypted with the correct public key. Depending on the service you want to use, this key can be downloaded using this command <strong>if you are uploading to the SDA</strong>:

Reviewer 2: While `<strong>` works, I prefer `**` for markdown.
```suggestion
The files that are uploaded to the SDA/BP services, need to be encrypted with the correct public key. Depending on the service you want to use, this key can be downloaded using this command **if you are uploading to the SDA**:
```

```bash
wget https://raw.githubusercontent.com/NBISweden/EGA-SE-user-docs/main/crypt4gh_key.pub
```

> or this command <strong>if you are uploading to Big Picture</strong>:

Reviewer 2: ```suggestion
or this command **if you are uploading to Big Picture**:
```

## Get dataset size

Before downloading a dataset or a specific file, the `sda-cli` tool allows for requesting the size of each file, as well as the whole dataset. In order to use this functionality, the tool expects as an argument a file containing the location of the files in the dataset. The argument can be one of the following:
1. a URL to the file containing the locations of the dataset files
2. a URL to a folder containing the `urls_list.txt` file with the locations of the dataset files
3. the path to a local file containing the locations of the dataset files.

Given this argument, the dataset size can be retrieved using the following command:
```bash
./sda-cli datasetsize <urls_file>
```
where `urls_file` as described above.

## Download

The SDA/BP archive enables for downloading files and datasets in a secure manner. That can be achieved using the `sda-cli` tool and the process consists of the following two steps

### Create keys

> In order to make sure that the files are downloaded from the archive in a secure manner, the user is supposed to create they key pair that the files will be encrypted with. The key pair can be created using the following command:

Reviewer 2: ```suggestion
In order to make sure that the files are downloaded from the archive in a secure manner, the user is supposed to create the key pair that the files will be encrypted with. The key pair can be created using the following command:
```

```bash
./sda-cli createKey <keypair_name>
```

> where `<keypair_name>` the names of the key files. This command will create two keys named `keypair_name.pub.pem` and `keypair_name.sec.pem`. The public key (`pub`) will be used for the encryption of the files, while the private one (`sec`) will be used in the decryption step below.

Reviewer 2: ```suggestion
where `<keypair_name>` is the the base name of the key files. This command will create two keys named `keypair_name.pub.pem` and `keypair_name.sec.pem`. The public key (`pub`) will be used for the encryption of the files, while the private one (`sec`) will be used in the decryption step below.
```

**NOTE:** Make sure to keep these keys safe. Losing the keys could lead to sensitive data leaks.

### Download file

The `sda-cli` tool allows for downloading file(s)/datasets. The whole dataset is stored in a file named `urls_list.txt`. Using this file or the URL where it is stored, it is possible to download the files. There are three different ways to pass the location of the file to the tool, similar to the [dataset size section](#get-dataset-size):
1. a URL to the file containing the locations of the dataset files
2. a URL to a folder containing the `urls_list.txt` file with the locations of the dataset files
3. the path to a local file containing the locations of the dataset files.

Given this argument, the dataset can be retrieved using the following command:
```bash
./sda-cli download <urls_file>
```
where `urls_file` as described above.

> The tool also allows for selecting a folder where the files will be downloading, using the `outdir` argument like:

Reviewer: ```suggestion
The tool also allows for selecting a folder where the files will be downloaded, using the `outdir` argument like:
```

```bash
./sda-cli download -outdir <outdir> <urls_file>
```

## Decrypt file

Given that the instructions in the [download section](#download) have been followed, the key pair and the data files should be stored in some location. The last step is to decrypt the files in order to access their content. That can be achieved using the following command:
```bash
./sda-cli decrypt -key <keypair_name>.sec.pem <file_to_decrypt>
```

> where `<keypair_name>.sec.pem` the private key created in the [relevant section](#create-keys) and `<file_to_decrypt>` one of the files downloaded following the instructions of the [download section](#download-file).

Reviewer: Maybe add here that multiple files can be decrypted as well.
